dockers/docker-orchagent/swssconfig.sh

```diff
@@ -39,7 +39,10 @@ rm -f /ready
 # Restore FDB and ARP table ASAP
 fast_reboot
 
-HWSKU=`sonic-cfggen -d -v "DEVICE_METADATA['localhost']['hwsku']"`
+# read SONiC immutable variables
+[ -f /etc/sonic/sonic-environment ] && . /etc/sonic/sonic-environment
+
+HWSKU=${HWSKU:-`sonic-cfggen -d -v "DEVICE_METADATA['localhost']['hwsku']"`}
```

Comment:
>sonic-cfggen [](start = 16, length = 12)

It is capture the intelligence into `sonic-cfggen`? Then you can minimize your code changes.

Reply:
Good point! however to reduce calling into python with python loading overhead. It is critical path and every msec would help here.

Reply:
Ultimately, I'd hope for eliminate such fallback logic once sonic environment file is guaranteed to be present.

Reply:
To make my proposal more concrete, you can have a bash function called `sonic-cfggen` which capture the intelligence and finally call executable `sonic-cfggen`.

The function could be setup in `/root/.bashrc`, or any other possible place. Maybe possible in rc.local.

I see the code change radius is so large, and the code readability decrease a lot. Ideally you fix one issue in one place. If not possible, there is some refractory work needed.

---
In reply to: [458361788](https://github.com/Azure/sonic-buildimage/pull/5010#discussion_r458361788) [](ancestors = 458361788)

Reply:
This would involve taking python sonic-cfggen out of the $PATH. or renaming it and redirect calls to it. However the bash script now needs to have logic to parse the command line and know if this is an environment variable or not. This will create another complex piece with possible errors.

The current proposal has the benefits of moving away from sonic-cfggen once the environment variable become mainstream. For those immutable properties, we do not need to go through cfggen. The current solution provide the fallback mechanism until all paths that are changing the environment var are changed to generate those variable. So, we do not need to go through cfggen or complex script.

Reply:
> This would involve taking python sonic-cfggen out of the $PATH.

This is not true. If there are function and binary with the same name, you can still call the binary by `$(which sonic-cfggen)`.

My intention is to simplify code changes, no intention to make changes complex.

---
In reply to: [458431281](https://github.com/Azure/sonic-buildimage/pull/5010#discussion_r458431281) [](ancestors = 458431281,458361788)

Reply:
Agree, simplicity is important. My target is to minimize any overhead in this paths as we are running tight in our CPU budget. The shim layer will create extra shell env and involves text processing for every call to sonic-cfggen. There are also different style when calling cfggen for specific var such as DEVICE_METADAT.localhost.platform, DEVICE_METADAT['localhost']['platform'] or, etc.  This would add more CPU overhead in the current path.